### PR TITLE
Fix Escaping bug

### LINKED ISSUE #931: Exclude not working with escaped characters
It appears there is a regression with Exclude and Remove tags not excluding escaped names. `<Compile Include=”a?b.my” Exclude=”a%27b.my” />` (or with a `<Compile Remove=”a%27b.my” />` still includes `a’b.my`. This worked with previous builds.

### PR DESCRIPTION
**Pull request by cdmihai**
Fixes #931

### CODE REVIEW

File: src/XMakeBuildEngine/UnitTestsPublicOM/Definition/ProjectItem_Tests.cs [hunk 1]
```diff
@@ -1912,6 +1998,128 @@ public void UpdateShouldUseCaseInsensitiveMatching()
...
+            get
+
+            {
+                var expectedMetadata = new[]
+                {
+                    new Dictionary<string, string> {{"m", "contents"}},
+                    new Dictionary<string, string> {{"m", "updated"}}
+                };
+
+                // escaped value matches and nonescaped value include
```
Comment (rainersigwald) [inline src/XMakeBuildEngine/UnitTestsPublicOM/Definition/ProjectItem_Tests.cs]: `/s/ and / a /g`?

**Comment by rainersigwald** [comment]
LGTM but go in with the other one first, then (maybe) rebase this onto it.

File: src/XMakeBuildEngine/UnitTestsPublicOM/Definition/ProjectItem_Tests.cs (hunk 1 above)
Comment (cdmihai) [inline src/XMakeBuildEngine/UnitTestsPublicOM/Definition/ProjectItem_Tests.cs]: More descriptive comments? :)

**Comment by cdmihai** [comment]
I rebased this on master after #934 got merged in. Good to go?

File: src/Shared/FileUtilities.cs [hunk 2]
```diff
@@ -470,16 +470,33 @@ internal static string GetFullPathNoThrow(string path)
...
         /// <summary>
+        /// Normalizes a path for path comparison
+        /// Does not throw IO exceptions. See <see cref="GetFullPathNoThrow(string, string)"/>
+        /// 
+        /// </summary>
+        internal static string NormalizePathForComparisonNoThrow(string path, string currentDirectory)
+        {
+            return GetFullPathNoThrow(path, currentDirectory)
+                .TrimEnd(Path.DirectorySeparatorChar)
+                .TrimEnd(Path.AltDirectorySeparatorChar);
```
Comment (rainersigwald) [inline src/Shared/FileUtilities.cs]: Do this in a single call (and does this one need to have the literal characters due to the forward/backslash thing?).

File: src/Shared/FileUtilities.cs [hunk 3]
```diff
@@ -470,16 +470,33 @@ internal static string GetFullPathNoThrow(string path)
         /// <returns></returns>
         internal static bool ComparePathsNoThrow(string first, string second, string currentDirectory)
         {
-            // todo: on xplat, figure out if file system is case sensitive or not
+            // TODO: assumption on file system case sensitivity: https://github.com/Microsoft/msbuild/issues/781
             var stringComparison = StringComparison.OrdinalIgnoreCase;
 
-            var firstFullPath = GetFullPathNoThrow(first, currentDirectory);
-            var secondFullPath = GetFullPathNoThrow(second, currentDirectory);
+            if (string.Equals(first, second, stringComparison))
```
Comment (rainersigwald) [inline src/Shared/FileUtilities.cs]: Is this just for perf?

File: src/Shared/UnitTests/ObjectModelHelpers.cs [hunk 4]
```diff
@@ -1414,5 +1414,31 @@ private static string[] SplitIntoLines(string content)
...
+                else
+                {
+                    ObjectModelHelpers.AssertItems(expectedInclude, new Project(projectFile).Items.ToList(), expectedMetadataPerItem);
+                }
+                
+            }
+            finally
+            {
+                ObjectModelHelpers.DeleteDirectory(root);
+                Directory.Delete(root);
```
Comment (rainersigwald) [inline src/Shared/UnitTests/ObjectModelHelpers.cs]: This second call seems redundant. Not true?

**Comment by rainersigwald** [comment]
:shipit:

File: src/Shared/FileUtilities.cs (hunk 2 above)
Comment (cdmihai) [inline src/Shared/FileUtilities.cs]: Ah, very true.

File: src/Shared/FileUtilities.cs (hunk 3 above)
Comment (cdmihai) [inline src/Shared/FileUtilities.cs]: Yes. Should I call it out with a comment?

Comment (rainersigwald) [inline src/Shared/FileUtilities.cs]: Yes, please.

File: src/Shared/UnitTests/ObjectModelHelpers.cs (hunk 4 above)
Comment (cdmihai) [inline src/Shared/UnitTests/ObjectModelHelpers.cs]: https://github.com/Microsoft/msbuild/blob/6d98ba349e6273ffa7b0cfdf518885ee6475cf11/src/Shared/UnitTests/ObjectModelHelpers.cs#L692-L694

Comment (rainersigwald) [inline src/Shared/UnitTests/ObjectModelHelpers.cs]: Oh, sure. 🤦 Carry on.

**Comment by cdmihai** [comment]
@dotnet-bot test this please

**Comment by cdmihai** [comment]
@dotnet-bot test OSX Build for CoreCLR please

**Pull request merged**